### DISCUSSION TITLE
3D Extruded shapes - start height/elevation

### DISCUSSION THREAD

**Discussion by MartinP-C** [visgl/deck.gl#6877] [Q&A] [answered]
Is it possible to define a start height for the extrusion, instead of always extruded from the ground level?

We wish to render a single shape, but cut up in to layers. So would create `n` shapes at, for example, 0 - 10m, 10m - 20m, 20 - `n`m

Or failing that, if given a single polygon with a color for each layer, how could that be rendered in deck.gl?

**Comment by ibgreen** [comment]
Perhaps you can use the `modelMatrix` prop with a simple z translation matrix to raise the ground level for each layer?

**Comment by Pessimistress** [comment, marked as answer]
It's not clear which layer you are using when you say "shapes." All official layers support 3D positions. For example, if you are using `PolygonLayer`, see https://deck.gl/docs/api-reference/layers/polygon-layer#getpolygon

**Reply by MartinP-C** [reply]
I'm not using any specific layer at the moment, whichever one allows setting the start elevation of a shape/polygon! :)

You are right. I missed the part that mentions the polygon layers will accept `z` coords.
I can create a geojson layer where the data gives a flat polygon at `n` elevation, then another feature with the same `x, y`s at `n+10` elevation. Extruding both gives the stacked appearance:
![image](https://user-images.githubusercontent.com/8778016/166643117-2efb70e1-6dcc-41a6-841b-0ee9fd05f10b.png)

Now I am wondering, can I specify the z coords within a polygon layer, and have the shape "filled in"?
I suspect this is moving into 3D / WebGL territory that is out of scope for a deck.gl discussion..
```
new PolygonLayer({
    id: "poly",
    opacity: 0.8,
    stroked: false,
    filled: true,
    extruded: false,
    wireframe: true,
    getFillColor: [255, 0, 0],
    data: [
    {
        coords: [
        [
            [-123.15, 49.26, 0],
            [-123.20, 49.22, 0],
            [-123.04, 49.20, 0],
            [-123.15, 49.26, 0],
            
            [-123.15, 49.26, 400],
            [-123.20, 49.22, 400],
            [-123.04, 49.20, 400],

            [-123.15, 49.26, 0],
        ],
        ],
    },
    {
        coords: [
        [
            [-123.15, 49.26, 400],
            [-123.20, 49.22, 400],
            [-123.04, 49.20, 400],
            [-123.15, 49.26, 400],
            
            [-123.15, 49.26, 800],
            [-123.20, 49.22, 800],
            [-123.04, 49.20, 800],

            [-123.15, 49.26, 400],
        ],
        ],
    },
    ],
    getPolygon: (d) => d.coords,
}),
```

This set of coordinates ends up giving the three planes. Can a "filled in" 3D polygon be specified without using extrude?

![image](https://user-images.githubusercontent.com/8778016/166644715-804ca69f-287d-44cb-82b1-a1510182f3d6.png)

**Reply by Pessimistress** [reply]
I don't understand the question here. Does the layer not give you the behavior that you expect?

**Reply by MartinP-C** [reply]
Either not the behaviour I expect, or I have created the coordinates wrong.

How do I render a 3D shape using only coordinates, not using `getElevation`?

**Reply by Pessimistress** [reply]
The PolygonLayer, as the name suggests, renders a surface ("polygon") if `extruded: false`. If you want to define a free form 3D object you can look at the SimpleMeshLayer.
